### PR TITLE
[IL-59] implement critical user registration api tests

### PR DESCRIPTION
**Pull request by yuliahar**
**Task's name:**
API tests for user registration

**Changes:**
1. Add critical api tests for user registration endpoint:
* register with non-unique email
* register with missing mandatory field

2. Fix the check_mapping_api_to_db method in [framework/asserts/registration_asserts.py]: add when key_api is "password," the code now hashes the password from the API request and compares it to the hashed password in the database.

3. Add .gitignore
4. Add bcrypt lib to requirements.txt

⚠️ Please note: the following tests are skipped for now due to bug in API (wrong validation message)
* register with non-unique email
* register with missing mandatory field

### CODE REVIEW

File: framework/asserts/registration_asserts.py [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
-        api_request:    reference data, data from an API-request;
-        database_data:  compared data, data from database.
+        api_request:    reference data, data from an API request;
+        database_data:  compared data, data from the database (a dictionary).
     """
     fields_api_to_db = {
         "email": "email",
         "firstName": "first_name",
         "lastName": "last_name",
         "password": "password",
```
Comment (Sanya124) [inline framework/asserts/registration_asserts.py:16]: Checking the password at 19 line => need delete this line (`"password": "password",`)

File: tests/registration/test_registration.py [hunk 2]
```diff
@@ -1,5 +1,7 @@
+import pytest
 from allure import description, feature, link, step, title
-from hamcrest import assert_that, is_
+from allure_commons._allure import severity
```
Comment (Sanya124) [inline tests/registration/test_registration.py]: I think `severity` to add at 2 line

File: tests/registration/test_registration.py [hunk 3]
```diff
@@ -13,35 +15,136 @@
...
-    @title("User registration under minimum requirements")
-    @description(
-        "WHEN the user submits the minimum required data for registration, "
-        "THEN the information about the user is stored in the database"
-    )
-    def test_of_registration(self, postgres):
-        with step("Generation data for registration"):
-            email = generate_string(length=2, additional_characters=["@te.st"])
+    """
+    Test suite for testing the registration of a new user
```
Comment (Sanya124) [inline tests/registration/test_registration.py]: It is not necessary to add a description to the tests. Title and description should explain everything

File: tests/registration/test_registration.py [hunk 4]
```diff
@@ -13,35 +15,136 @@
...
+    @title("User registration under minimum requirements")
+    @description(
+        "WHEN the user submits the minimum required data for registration, "
+        "THEN the information about the user is stored in the database"
+    )
+    def test_of_registration(self, postgres):
+        """
+        Test of registration of user with minimum requirements
+        Args:
+            postgres: fixture for working with the database
```
Comment (Sanya124) [inline tests/registration/test_registration.py]: It is not necessary to add a description to the tests. Title and description should explain everything

File: tests/registration/test_registration.py [hunk 5]
```diff
@@ -13,35 +15,136 @@
...
+    )
+    def test_of_registration_required_fields(self, postgres, data):
+        """
+        Test of registration of user with missing required field
+        Args:
+            postgres: fixture for working with the database
+            data: tuple with field and expected error message
+        """
+        [field, expected_message] = data
+        self.user_to_register.pop(field)
```
Comment (Sanya124) [inline tests/registration/test_registration.py]: wrap in step

File: tests/registration/test_registration.py [hunk 6]
```diff
@@ -13,35 +15,136 @@
...
+            user_data = postgres.get_data_by_filter(
+                table="user_details", field="email", value=self.email
+            )
+
+        with step("Checking that new user with duplicate email has not been registered "):
+            assert_that(user_data, has_length(1))
+            check_mapping_api_to_db(api_request=self.user_to_register, database_data=user_data[0])
+
+    fields = [
+        # TODO: check the error message for the email field
```
Comment (Sanya124) [inline tests/registration/test_registration.py]: let's clarify first, and then describe how it should be and we can add `pytest.mark.xfail`

File: framework/asserts/registration_asserts.py (hunk 1 above)
Comment (yuliahar) [reply framework/asserts/registration_asserts.py:16]: Row 19 doesn't verify if the passwords match; it just hashes the new user's password using the database password and new user password as the salt. However, password comparison still occurs within a loop.

File: tests/registration/test_registration.py (hunk 2 above)
Comment (yuliahar) [reply tests/registration/test_registration.py]: Fixed

File: tests/registration/test_registration.py (hunk 3 above)
Comment (yuliahar) [reply tests/registration/test_registration.py]: Fixed

File: tests/registration/test_registration.py (hunk 4 above)
Comment (yuliahar) [reply tests/registration/test_registration.py]: Fixed

File: tests/registration/test_registration.py (hunk 5 above)
Comment (yuliahar) [reply tests/registration/test_registration.py]: Fixed

File: tests/registration/test_registration.py (hunk 6 above)
Comment (yuliahar) [reply tests/registration/test_registration.py]: Commented for now, waiting for response.

File: framework/asserts/registration_asserts.py (hunk 1 above)
Comment (Sanya124) [reply framework/asserts/registration_asserts.py:16]: I apologize. I didn't understand the situation ((

File: tests/registration/test_registration.py (hunk 6 above)
Comment (Sanya124) [reply tests/registration/test_registration.py]: @yuliahar if you take responsibility that you won't forget and you will definitely fix it in the near future, then let's merge :)

**Review by Sanya124 — APPROVED** [review]

File: tests/registration/test_registration.py (hunk 6 above)
Comment (yuliahar) [reply tests/registration/test_registration.py]: agreed :)

**Review by Sunagatov — APPROVED** [review]

**Pull request merged**